ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java

```diff
@@ -216,21 +290,102 @@ private String readApplicationId(Process exec) {
+      String releaseCmd = exportCmd + " && " + releaseCommand;
       String[] cmd = {"/bin/sh", "-c", releaseCmd};
-      LOG.info("Releasing flink optimizer using command:" + releaseCmd);
-      Runtime.getRuntime().exec(cmd);
+      LOG.info("Releasing flink optimizer using command: " + releaseCmd);
+      Process exec = Runtime.getRuntime().exec(cmd);
+      fetchCommandOutput(exec, null);
     } catch (IOException e) {
       throw new UncheckedIOException("Failed to release flink optimizer.", e);
     }
   }
+
+  private String buildReleaseYarnCommand(Resource resource) {
+    Preconditions.checkArgument(resource.getProperties().containsKey(YARN_APPLICATION_ID_PROPERTY),
+        "Cannot find {} from optimizer start up stats", YARN_APPLICATION_ID_PROPERTY);
+    String applicationId = resource.getProperties().get(YARN_APPLICATION_ID_PROPERTY);
+    String options = "-Dyarn.application.id=" + applicationId;
+
+    Preconditions.checkArgument(resource.getProperties().containsKey(FlinkOptimizer.JOB_ID_PROPERTY),
+        "Cannot find {} from optimizer properties", FlinkOptimizer.JOB_ID_PROPERTY);
+    String jobId = resource.getProperties().get(FlinkOptimizer.JOB_ID_PROPERTY);
+    return String.format("%s/bin/flink cancel -t %s %s %s",
+        flinkHome, target.getValue(), options, jobId);
+  }
+
+  private String buildReleaseKubernetesCommand(Resource resource) {
+    Preconditions.checkArgument(resource.getProperties().containsKey(KUBERNETES_CLUSTER_ID_PROPERTY),
+        "Cannot find {} from optimizer start up stats.", KUBERNETES_CLUSTER_ID_PROPERTY);
+    String clusterId = resource.getProperties().get(KUBERNETES_CLUSTER_ID_PROPERTY);
+
+    FlinkConf conf = FlinkConf.buildFor(loadFlinkConfig(), getContainerProperties())
+        .withGroupProperties(resource.getProperties()).build();
+    conf.putToOptions(FlinkConf.PropertyKeys.KUBERNETES_CLUSTER_ID, clusterId);
+
+    String options = conf.toCliOptions();
+
+    // job-manager rest api can't be visited from k8s cluster outside, so using kubernetes-session command to kill it
+    // echo 'stop' | $FLINK_HOME/bin/kubertnetes-session.sh
+    return String.format(" echo 'stop' | %s/bin/kubernetes-session.sh %s -Dexecution.attached=true",
+        this.flinkHome, options);
+  }
+
+  private String getFlinkHome() {
+    String flinkHome = PropertyUtil.getRequiredNotNull(getContainerProperties(), FLINK_HOME_PROPERTY);
+    return flinkHome.replaceAll("/$", "");
+  }
+
+  private String getFlinkConfDir() {
+    String flinkConfDir = getContainerProperties().get(PropertyNames.EXPORT_PROPERTY_PREFIX + ENV_FLINK_CONF_DIR);
+    if (StringUtils.isNotEmpty(flinkConfDir)) {
+      return flinkConfDir;
+    }
+    return this.flinkHome + FLINK_CONFIG_PATH;
+  }
+
+  private String kubernetesClusterId(Resource resource) {
+    return "amoro-optimizer-" + resource.getResourceId();
+  }
+
+  private enum Target {
+    YarnPreJob("yarn-per-job", false),
```

Comment:
Upper case and under line?

Reply:
fixed

Reply:
I mean is that enum types are defined using uppercase and underline.
```
    YARN_PER_JOB("yarn-per-job", false),
    YARN_APPLICATION("yarn-application", true),
    KUBERNETES_APPLICATION("kubernetes-application", true);
```

Reply:
fixed